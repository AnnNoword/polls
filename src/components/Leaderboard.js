import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Nav from './Nav';

const Leaderboard = ({ users }) => {
    const tableHeadings = ['Users', 'Answered', 'Created'];
    return (
        <div>
            <Nav />
            <div className=" max-w-xl mx-auto shadow-lg rounded-lg border p-4 flex flex-col">
                <h1 className="text-sky-800 text-3xl font-bold text-center my-4">Leaderboard</h1>
                <div className="overflow-x-auto sm:-mx-6 lg:-mx-8">
                    <div className="py-2 inline-block min-w-full sm:px-6 lg:px-8">
                        <div className="overflow-hidden">
                            <table className="min-w-full">
                                <thead className="border-b">
                                <tr>
                                    {
                                        tableHeadings.map(h => (
                                            <th key={h.toLowerCase()} scope="col" className="text-sm font-medium text-gray-900 px-6 py-4 text-left">
                                                {h}
                                            </th>
                                        ))
                                    }
                                </tr>
                                </thead>
                                <tbody>
                                {
                                    users.map((user) => {
                                        const questionsAnswered = Object.keys(user.answers).length,
                                            questionsAsked = user.questions.length;
                                        return (
                                            <tr key={user.id} className="border-b">
                                                <td className="flex text-sm text-gray-900  px-6 py-4 whitespace-nowrap">
                                                    <img src={user.avatarURL} alt={user.name} className="rounded-full h-16 w-16" />
                                                    <div className="py-4 pl-4">
                                                        <h6 className="font-semibold">{user.name}</h6>
                                                        <i>{user.id}</i>
                                                    </div>
                                                </td>
                                                <td className="text-center text-sm text-gray-900  px-6 py-4 whitespace-nowrap">
                                                    {questionsAnswered}
                                                </td>
                                                <td className="text-center text-sm text-gray-900  px-6 py-4 whitespace-nowrap">
                                                    {questionsAsked}
                                                </td>
                                            </tr>
                                        );
                                    })
                                }
                                </tbody>
                            </table>
                        </div>
                    </div>
                </div>
            </div>
        </div>
    );
};

Leaderboard.propTypes = {
    users: PropTypes.array.isRequired
};

const mapStateToProps = ({ users }) => {
    const usersList = Object.values(users);
    return {
        users: usersList.sort((a, b) => (Object.keys(b.answers).length + b.questions.length) - (Object.keys(a.answers).length + a.questions.length))
    };
};

export default connect(mapStateToProps)(Leaderboard);
