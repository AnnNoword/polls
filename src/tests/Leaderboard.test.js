/* eslint-disable */
import { render, screen } from '@testing-library/react';
import Leaderboard from '../components/Leaderboard';
import { legacy_createStore as createStore } from 'redux';
import reducer from '../reducers';
import middleware from '../middleware';
import { Provider } from 'react-redux';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

const store = createStore(reducer, middleware);

test('render Leaderboard component', () => {
    render(
        <MemoryRouter>
            <Provider store={store}>
                <Leaderboard />
            </Provider>
        </MemoryRouter>
    );
    screen.debug();
});

describe('Leaderboard', () => {
    const component = render(
        <MemoryRouter>
            <Provider store={store}>
                <Leaderboard />
            </Provider>
        </MemoryRouter>);
    expect(component).toMatchSnapshot();
});
