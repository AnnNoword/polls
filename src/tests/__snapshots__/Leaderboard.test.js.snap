// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="flex justify-between space-x-4 border-b mb-8 items-center pb-4 mt-8"
        >
          <nav
            class="space-x-4 py-1"
          >
            <a
              aria-current="page"
              class="bg-slate-200 rounded-lg px-3 py-2 text-slate-700 font-medium"
              href="/"
            >
              Home
            </a>
            <a
              class="link rounded-lg px-3 py-2 text-slate-700 font-medium hover:bg-slate-100 hover:text-slate-900"
              href="/leaderboard"
            >
              Leaderboard
            </a>
            <a
              class="link rounded-lg px-3 py-2 text-slate-700 font-medium hover:bg-slate-100 hover:text-slate-900"
              href="/add"
            >
              New
            </a>
          </nav>
          <nav
            class="justify-end ml-8"
          />
        </div>
        <div
          class=" max-w-xl mx-auto shadow-lg rounded-lg border px-4 py-4 flex flex-col"
        >
          <h1
            class="text-sky-800 text-3xl font-bold text-center mt-4 mb-4"
          >
            Leaderboard
          </h1>
          <div
            class="overflow-x-auto sm:-mx-6 lg:-mx-8"
          >
            <div
              class="py-2 inline-block min-w-full sm:px-6 lg:px-8"
            >
              <div
                class="overflow-hidden"
              >
                <table
                  class="min-w-full"
                >
                  <thead
                    class="border-b"
                  >
                    <tr>
                      <th
                        class="text-sm font-medium text-gray-900 px-6 py-4 text-left"
                        scope="col"
                      >
                        Users
                      </th>
                      <th
                        class="text-sm font-medium text-gray-900 px-6 py-4 text-left"
                        scope="col"
                      >
                        Answered
                      </th>
                      <th
                        class="text-sm font-medium text-gray-900 px-6 py-4 text-left"
                        scope="col"
                      >
                        Created
                      </th>
                    </tr>
                  </thead>
                  <tbody />
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="flex justify-between space-x-4 border-b mb-8 items-center pb-4 mt-8"
      >
        <nav
          class="space-x-4 py-1"
        >
          <a
            aria-current="page"
            class="bg-slate-200 rounded-lg px-3 py-2 text-slate-700 font-medium"
            href="/"
          >
            Home
          </a>
          <a
            class="link rounded-lg px-3 py-2 text-slate-700 font-medium hover:bg-slate-100 hover:text-slate-900"
            href="/leaderboard"
          >
            Leaderboard
          </a>
          <a
            class="link rounded-lg px-3 py-2 text-slate-700 font-medium hover:bg-slate-100 hover:text-slate-900"
            href="/add"
          >
            New
          </a>
        </nav>
        <nav
          class="justify-end ml-8"
        />
      </div>
      <div
        class=" max-w-xl mx-auto shadow-lg rounded-lg border px-4 py-4 flex flex-col"
      >
        <h1
          class="text-sky-800 text-3xl font-bold text-center mt-4 mb-4"
        >
          Leaderboard
        </h1>
        <div
          class="overflow-x-auto sm:-mx-6 lg:-mx-8"
        >
          <div
            class="py-2 inline-block min-w-full sm:px-6 lg:px-8"
          >
            <div
              class="overflow-hidden"
            >
              <table
                class="min-w-full"
              >
                <thead
                  class="border-b"
                >
                  <tr>
                    <th
                      class="text-sm font-medium text-gray-900 px-6 py-4 text-left"
                      scope="col"
                    >
                      Users
                    </th>
                    <th
                      class="text-sm font-medium text-gray-900 px-6 py-4 text-left"
                      scope="col"
                    >
                      Answered
                    </th>
                    <th
                      class="text-sm font-medium text-gray-900 px-6 py-4 text-left"
                      scope="col"
                    >
                      Created
                    </th>
                  </tr>
                </thead>
                <tbody />
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
